### DISCUSSION TITLE
Noob looking for help

### DISCUSSION THREAD

**Discussion by JohnDepot** [ImageMagick/ImageMagick#2434] [Help]
Brand new to ImageMagick and I am running head first into the learning curve!

I have a set of images (.bmp) that contains a description _ Height (px) and Width (px) (.jpg version included below)

What I need to do is to run a script that will take the original image "Square_160x160" and make multiple colors of the same image, adding the color to the file name "Green_Square_160x160" "Blue_Square_160x160" "Yellow_Square_160x160" etc.

So far Ive been able to use the example script to make a Gray version, but cant make it a color.

convert Square_160x160.bmp -colorspace RGB -colorspace Gray Gray_Square_160x160.bmp

Any help would be appreciated, thank you for your time!

![Square_160x160](https://user-images.githubusercontent.com/69867563/90548751-6c17e400-e15b-11ea-90ca-6322a7bba610.jpg)

**Comment by fmw42** [comment]
Here is one way. Make white into transparent. Then convert to grayscale. Then use +level-colors to add black to blue shading, then change the transparency to white again.

`convert button.jpg -fuzz 15% -transparent white -colorspace gray +level-colors black,blue -background white -compose over -flatten result.png`

![result](https://user-images.githubusercontent.com/13458218/90551178-d160da80-e145-11ea-8365-0fb48804711c.png)

**Reply by JohnDepot** [reply]
fmw42:
Looks great, but im trying to keep a white border and alpha, let me try to get a comparison picture up here:

So running with the code example:
convert S

> quare_160x160.bmp -fuzz 15% -transparent white -colorspace gray +level-colors black,blue -background white -compose over -flatten Blue_Square_160x160.bmp
> convert Square_160x160.bmp -fuzz 15% -transparent white -colorspace gray +level-colors black,green -background white -compose over -flatten Green_Square_160x160.bmp
> convert Square_160x160.bmp -fuzz 15% -transparent white -colorspace gray +level-colors black,red -background white -compose over -flatten Red_Square_160x160.bmp
> convert Square_160x160.bmp -fuzz 15% -transparent white -colorspace gray +level-colors black,yellow -background white -compose over -flatten Yellow_Square_160x160.bmp
> 

What I get from the code:

![4 color w code](https://user-images.githubusercontent.com/69867563/90553126-dc296880-e161-11ea-9cb7-32cfb2986f8b.jpg)

What I am looking for:

![4 color manual](https://user-images.githubusercontent.com/69867563/90553129-de8bc280-e161-11ea-8fd3-b1a3f716e54a.jpg)

**Comment by JohnDepot** [comment]
This is the base bmp file
[Square_160x160.zip](https://github.com/ImageMagick/ImageMagick/files/5092317/Square_160x160.zip)

**Reply by fmw42** [reply]
This is producing an invalid bmp. Same below as well

**Comment by JohnDepot** [comment]
[Icon color changer.zip](https://github.com/ImageMagick/ImageMagick/files/5092327/Icon.color.changer.zip)
example icons, the symbol will be hard white, just need to change the background of the button

**Comment by fmw42** [comment]
This approach preserves black and white and your transparency. It creates a color map. Then uses -clut to apply that to the grayscale version of the image. I saved as PNG since this forum does not allow uploads of BMP. But you can save to BMP if you want.

`convert Off_193x193.bmp -colorspace gray \( xc:black xc:blue xc:white +append -resize 1024x1 \) -clut result.png`

![result](https://user-images.githubusercontent.com/13458218/90555379-1daf1900-e14c-11ea-9cd5-015abf153bbe.png)

**Comment by JohnDepot** [comment]
That picture looks perfect but I'm unable to reproduce with the code snip without error:

C:\Users\***\Desktop\Vault\Icon color changer>convert Off_193x193.bmp -colorspace gray \( xc:black xc:blue xc:white +append -resize 1024x1 \) -clut result.png
convert: unable to open image '\(': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/562.
convert: unable to open image '\)': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: unable to open image '\)': No such file or directory @ error/blob.c/OpenBlob/3537.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/562.
convert: image sequence is required `-clut' @ error/mogrify.c/MogrifyImageList/7939.
convert: no images defined `result.png' @ error/convert.c/ConvertImageCommand/3285.

Press any key to continue . . .

Did I miss something?

**Reply by fmw42** [reply]
Windows syntax is slightly different. Remove the \ on the parentheses. Are you using Windows 10 Unix? If so, then my \ escapes are needed. Your command looks ok to me. Try adding -size 1x1 before xc:black. Be sure you have spaces on each side of the parentheses. What is your IM version? If using IM 7, then change convert to magick.

**Reply by JohnDepot** [reply]
Im just using Windows cmd to test and making .bat files for multi icon scripting. Using IM 7.0.10

magick Off_193x193.bmp -colorspace gray ( -size 1x1 xc:black xc:blue xc:white +append -resize 1024x1 ) -clut Blue_Off_193x193.bmp
magick Off_193x193.bmp -colorspace gray ( -size 1x1 xc:black xc:green xc:white +append -resize 1024x1 ) -clut Green_Off_193x193.bmp
magick Off_193x193.bmp -colorspace gray ( -size 1x1 xc:black xc:red xc:white +append -resize 1024x1 ) -clut Red_Off_193x193.bmp
magick Off_193x193.bmp -colorspace gray ( -size 1x1 xc:black xc:yellow xc:white +append -resize 1024x1 ) -clut Yellow_Off_193x193.bmp

Results attached and they look really good! Now I need to figure out what all this code actually does to see if i can try to keep the grays and resolution, just change the hue.
![image](https://user-images.githubusercontent.com/69867563/90563560-084ce580-e172-11ea-94fb-de830b9b8df8.png)
![image](https://user-images.githubusercontent.com/69867563/90563591-169b0180-e172-11ea-8345-49b2f16d277d.png)


[Off Icon color changer.zip](https://github.com/ImageMagick/ImageMagick/files/5092718/Off.Icon.color.changer.zip)

**Reply by fmw42** [reply]
See -modulate

`convert Off_193x193.bmp -modulate 100,100,33 result.png`

![x](https://user-images.githubusercontent.com/13458218/90565875-68389180-e15c-11ea-83c7-4d62184220cf.png)

**Reply by fmw42** [reply]
Note that modulate only shifts hues. You cannot just say what hue you want. But if you start with a red icon/button, then the shift and desired hue are the same. Also note that modulate works with values in the range 0 to 200 rather than 0 to 360. So the conversion from hue shift to what modulate needs is 200*hue/360, where hue is the shift in degrees.

**Comment by GeeMack** [comment]
To get a simple hue change with ImageMagick try "-modulate" as fmw42 mentioned above. Another idea might be to make the image gray then apply color with "-tint". Setting the color and using it as a filename can be easily done with this approach as long as the color name is understood by IM.

    set buttoncolor=purple
    magick Square_160x160.bmp -modulate 100,0 -fill %buttoncolor% -tint 100 Button_%buttoncolor%.png

![Button_purple](https://user-images.githubusercontent.com/65475279/90579158-c2504b80-e18a-11ea-952b-37dc49db8f0e.png)

**Comment by JohnDepot** [comment]
Thanks all for the help! Last minute a friend suggested using another program, ImBatch which was able to do a batch color channel changes and save each new image with a formula based name. 

![ImBatch result](https://user-images.githubusercontent.com/69867563/90645297-cbc9ca00-e203-11ea-86f4-dd3203fa916e.png)

Thank you all for your time and effort!
